src/snapred/backend/recipe/algorithm/GroupByGroupCalibration.py

```diff
@@ -0,0 +1,257 @@
+import json
+from typing import Dict, List
+
+from mantid.api import AlgorithmFactory, PythonAlgorithm
+from mantid.kernel import Direction
+
+from snapred.backend.dao.DiffractionCalibrationIngredients import DiffractionCalibrationIngredients
+from snapred.backend.recipe.algorithm.MantidSnapper import MantidSnapper
+
+name = "GroupByGroupCalibration"
+
+
+class GroupByGroupCalibration(PythonAlgorithm):
+    def PyInit(self):
+        # declare properties
+        self.declareProperty("Ingredients", defaultValue="", direction=Direction.Input)  # noqa: F821
+        self.declareProperty("InputWorkspace", defaultValue="", direction=Direction.Input)
+        self.declareProperty("PreviousCalibrationTable", defaultValue="", direction=Direction.Input)
+        self.declareProperty("OutputWorkspace", defaultValue="", direction=Direction.Output)
+        self.declareProperty("FinalCalibrationTable", defaultValue="", direction=Direction.Output)
+        self.setRethrows(True)
+        self.mantidSnapper = MantidSnapper(self, name)
+
+    def chopIngredients(self, ingredients):
+        """Receive the ingredients from the recipe, and exctract the needed pieces for this algorithm."""
+        from datetime import date
+
+        """Receive the ingredients from the recipe, and exctract the needed pieces for this algorithm."""
+        self.runNumber: str = ingredients.runConfig.runNumber
+        self.ipts: str = ingredients.runConfig.IPTS
+        self.rawDataPath: str = self.ipts + "shared/lite/SNAP_{}.lite.nxs.h5".format(ingredients.runConfig.runNumber)
+
+        # TODO setup for SNAPLite
+        self.isLite = False
+        self.stateFolder: str = ingredients.calPath
+
+        # set output filename
+        self.outputFilename: str
+        if self.isLite:
+            self.outputFilename = (
+                f'{self.stateFolder}SNAP{self.runNumber}_calib_geom_{date.today().strftime("%Y%m%d")}.lite.h5'
+            )
+        else:
+            self.outputFilename = (
+                f'{self.stateFolder}SNAP{self.runNumber}_calib_geom_{date.today().strftime("%Y%m%d")}.h5'
+            )
+
+        # fdrom the instrument state, read the overall min/max TOF
+        self.TOFMin: float = ingredients.instrumentState.particleBounds.tof.minimum
+        self.TOFMax: float = ingredients.instrumentState.particleBounds.tof.maximum
+        self.TOFBin: float = -0.001
```

Comment:
A comment would help as to why -0.001 value was chosen and why it's negative.

Reply:
It is negative as a flag to `Rebin` to use logarithmic binning.  I'm not sure why 0.001, except that it was hard-coded in Malcolm's script and there does not seem to be any other data object with it.  @mguthriem

Reply:
The choice of 0.001 is somewhat arbitrary and should be considered preliminary. It should also not be hard-coded, but should be pulled out. I think it probably makes sense to just store this alongside the other values as particleBounds.tof.binning